lib/elixir/lib/stream.ex

```diff
@@ -1021,25 +1024,28 @@ defmodule Stream do
     end
   end
 
-  defp do_zip([{fun, fun_acc} | t], acc, callback, list, buffer) do
+  defp do_zip([{fun, fun_acc} | t], acc, callback, yielded_elems, buffer) do
     case fun.({:cont, fun_acc}) do
       {:suspended, [i | fun_acc], fun} ->
-        do_zip(t, acc, callback, [i | list], [{fun, fun_acc} | buffer])
+        do_zip(t, acc, callback, [i | yielded_elems], [{fun, fun_acc} | buffer])
       {_, _} ->
         do_zip_close(:lists.reverse(buffer, t))
         {:done, acc}
     end
   end
 
-  defp do_zip([], acc, callback, list, buffer) do
-    zipped = List.to_tuple(:lists.reverse(list))
+  defp do_zip([], acc, callback, yielded_elems, buffer) do
+    # Here, "list" is the (reversed) list of results of this step of the
+    # enumeration (i.e., if we were zipping three streams, "list" would be a
+    # three-elements list of the elements currently yielded by each stream). We
+    # reverse this to fix the order and convert to tuple since the result of
+    # zipping n streams is a list of n-elements tuples.
+    zipped = List.to_tuple(:lists.reverse(yielded_elems))
     {:next, :lists.reverse(buffer), callback.(zipped, acc)}
   end
 
-  defp do_zip_close([]), do: :ok
-  defp do_zip_close([{fun, acc} | t]) do
-    fun.({:halt, acc})
-    do_zip_close(t)
```

Comment:
I replaced this with `:lists.foreach/2` since we were reimplementing exactly that, but with `foreach` the intent is crystal clear. I benchmarked this and there appears to be no performance regression (I wasn't expecting any but ¯\_(ツ)_/¯).